### PR TITLE
Scenes: Add new row and copy/paste functionalities

### PR DESCRIPTION
**Pull request by mdvictor**
Migrates the `Add row` and `copy/paste` functionalities in scenes. Part of https://github.com/grafana/grafana/issues/82316

TODO:

- [x] Tests
- [ ] ~Copy/Paste lib panels aswell~ -- added as a separate task in the milestone since there are dependencies on this one

### CODE REVIEW

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 1]
```diff
@@ -141,6 +149,7 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
       editable: true,
       body: state.body ?? new SceneFlexLayout({ children: [] }),
       links: state.links ?? [],
+      copiedPanel: store.exists(LS_PANEL_COPY_KEY),
```
Comment (torkelo) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:152]: not sure why we need this boolean flag in state (when it exits in local storage)

File: public/app/features/dashboard-scene/scene/DashboardScene.tsx [hunk 2]
```diff
@@ -423,26 +431,46 @@ export class DashboardScene extends SceneObjectBase<DashboardSceneState> {
...
-  public addPanel(vizPanel: VizPanel): void {
+  public addRow(row: SceneGridRow) {
     if (!(this.state.body instanceof SceneGridLayout)) {
       throw new Error('Trying to add a panel in a layout that is not SceneGridLayout');
     }
 
     const sceneGridLayout = this.state.body;
 
-    // move all gridItems below the new one
-    for (const child of sceneGridLayout.state.children) {
```
Comment (mdvictor) [inline public/app/features/dashboard-scene/scene/DashboardScene.tsx:434]: I realised now that grid cell sorting is done entirely in scenes and we don't need to worry about it here at all so I removed all logic for panel offsetting when something new is added

Comment (mdvictor) [reply public/app/features/dashboard-scene/scene/DashboardScene.tsx:434]: @ivanortegaalba Heads-up! I know we talked a bit about this in the other [PR](https://github.com/grafana/grafana/pull/83145).

File: public/app/features/dashboard-scene/utils/utils.ts [hunk 3]
```diff
@@ -200,35 +200,44 @@ export function isPanelClone(key: string) {
   return key.includes('clone');
 }
 
-export function getNextPanelId(dashboard: DashboardScene) {
+export function getNextPanelId(dashboard: DashboardScene): number {
```
Comment (mdvictor) [inline public/app/features/dashboard-scene/utils/utils.ts:203]: Wondering if this would sit better in `dashboardSceneGraph` ?

Comment (ivanortegaalba) [reply public/app/features/dashboard-scene/utils/utils.ts:203]: I'd say yes! 👍

File: public/app/features/dashboard-scene/utils/utils.test.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
+import { SceneGridItem, SceneGridLayout, SceneGridRow, SceneTimeRange, VizPanel } from '@grafana/scenes';
+
+import { DashboardControls } from '../scene/DashboardControls';
+import { DashboardScene, DashboardSceneState } from '../scene/DashboardScene';
+import { LibraryVizPanel } from '../scene/LibraryVizPanel';
+
+import { getNextPanelId } from './utils';
+
+describe('utils', () => {
+  describe('getNextPanelId', () => {
```
Comment (ivanortegaalba) [inline public/app/features/dashboard-scene/utils/utils.test.ts:10]: 💯

**Comment by mdvictor** [comment]
This got way bigger than I expected with all the import from lib modifications so I split that into a different [PR](https://github.com/grafana/grafana/pull/83498) that builds on top of this

**Review by ivanortegaalba — APPROVED** [review]
I tested all functionalities. All looks good to me! 💯 

✅ Add multiple rows, move components around
✅ The rows are properly positioned
✅ I copied/pasted multiple types of panels, and all of them keep everything working

The rows are missing the controls, but I guess this is coming in the following PR. Good job!

No scenes row
Captura de pantalla 2024-02-27 a las 23 20 20

Scenes row
Captura de pantalla 2024-02-27 a las 23 20 07

**Comment by mdvictor** [comment]
@ivanortegaalba yes, row controls are coming in a future PR

**Pull request merged**